File: Algorithms/Backtracking/79. Word Search.py
class Solution(object):
    def exist(self, board, word):
		def backtrack(row, col, inx):
			if(inx == len(word)):
				return True
			if(row < 0 or row == rowL or col < 0 or col == colL or board[row][col] != word[inx]):
				return False
			c = board[row][col]
			board[row][col] = '#'
			res = backtrack(row - 1, col, inx + 1) or backtrack(row + 1, col, inx + 1) or backtrack(row, col + 1, inx + 1) or backtrack(row, col - 1, inx + 1)
			board[row][col] = c
			return res 
			
		if(not board):
			return not word
		rowL, colL = len(board), len(board[0])
		for row in range(rowL):
			for col in range(colL):
				if(backtrack(row, col, 0)):
					return True
		return False

File: Algorithms/Backtracking/79. Word Search.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
		if(board.empty()) return false;
		char c = word[0];
		m = board.size();
		n = board[0].size();
		for(int i = 0; i < m; i++)
		{
			for(int j = 0; j < n; j++)
			{
				if(board[i][j] == c && check(board, word, 0, i, j)) return true;
			}
		}
		return false;
    }
private:
	int m, n;
	bool check(vector<vector<char>>& board, string& word, int inx, int r, int c) {
		if(word[inx] != board[r][c]) return false;	
		if(inx == word.size() - 1) return true;
		board[r][c] = '$';
		if(r < m - 1 && check(board, word, inx + 1, r + 1, c))
		{
			board[r][c] = word[inx];
			return true;
		}
		if(r > 0 && check(board, word, inx + 1, r - 1, c))
		{
			board[r][c] = word[inx];
			return true;
		}
		if(c < n - 1 && check(board, word, inx + 1, r, c + 1))
		{
			board[r][c] = word[inx];
			return true;
		}
		if(c > 0 && check(board, word, inx + 1, r, c - 1))
		{
			board[r][c] = word[inx];
			return true;
		}
		board[r][c] = word[inx];
		return false;
	}
};
